crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs

```diff
@@ -554,11 +617,94 @@ impl GroupSessionManager {
         // needed because each encryption step will mutate the Olm session,
         // ratcheting its state forward.
         for result in join_all(tasks).await {
-            let used_sessions: OlmResult<Vec<Session>> = result.expect("Encryption task panicked");
+            let result: OlmResult<(Vec<Session>, BTreeMap<OwnedUserId, Vec<OwnedDeviceId>>)> =
+                result.expect("Encryption task panicked");
 
-            changes.sessions.extend(used_sessions?);
+            let (used_sessions, failed_no_olm) = result?;
+            changes.sessions.extend(used_sessions);
+
+            no_olm.extend(failed_no_olm.into_iter())
         }
 
+        // Handle unverified and blacklisted withhelds
+        let mut flat_with_code: Vec<(OwnedUserId, OwnedDeviceId, WithheldCode)> = withheld_devices
+            .into_iter()
+            .flat_map(|(_, list)| {
+                list.into_iter().map(|(d, c)| (d.user_id().to_owned(), d.device_id().to_owned(), c))
+            })
+            .collect();
+
+        no_olm.iter().for_each(|(u, list)| {
```

Comment:
Again, why doesn't the `encrypt_request()` method give us the withheld devices in the correct format, then this becomes a simple operation of merging two `BTreeMaps`.

Reply:
refactored